[1단계 - 장바구니] 상추(이상희) 미션 제출합니다.

sanghee01 opened the pull request:
안녕하세요 도밥! 상추🥬입니다

레벨1 점뭐먹 미션 때 도밥을 통해 정말 다방면으로 많은 도움을 받았었는데 
이렇게 또 마지막 미션으로 만나뵙게 돼서 영광이예요 😆 

아직 부족한 점이 많지만 이번에도 잘 부탁드립니다.
바쁘신 와중에도 시간 내어 리뷰해주셔서 감사합니다!

## 📦 장바구니 미션

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

### 1단계

- 클라이언트 상태를 효과적으로 모델링하고 관리할 수 있다.
- Jest, React Testing Library(RTL)를 활용하여 주요 기능에 대한 테스트를 작성할 수 있다.

## 🕵️ 셀프 리뷰(Self-Review)

### 제출 전 체크 리스트

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] RTL 테스트 케이스를 모두 작성했나요?
- [x] 배포한 데모 페이지에 정상적으로 접근할 수 있나요?

  - 배포 링크 기입: https://sanghee01.github.io/react-shopping-cart/

- [x] 리뷰어가 장바구니 추가를 쉽게 할 수 있도록 Curl 명령어를 DM 으로 전달해주세요.

  ```
  curl -X 'POST' \
  '<url>' \
  -H 'accept: */*' \
  -H 'Authorization: Basic <토큰>' \
  -H 'Content-Type: application/json' \
  -d '{
  "productId": 1,
  "quantity": 1
  }'
  ```

### 리뷰 요청 & 논의하고 싶은 내용

### 1) 상태 설계 의도

* **Context API로 관리할 상태** - `selectedCartItems` vs. `cartItems`

  * Context API를 통해 전역으로 관리할 원본 상태와, 이를 기반으로 파생 상태를 만들어 데이터를 관리하고자 했습니다. 그 과정에서 Context API로 관리할 상태를 어떤 것으로 설정할지 고민이 있었습니다. 즉, 장바구니에서 선택된 상품 배열(`selectedCartItems`)이냐, 장바구니에 있는 모든 상품 배열(`cartItems`)이냐였습니다.
  * 각각 다음과 같은 장단점이 있어 고민하게 되었습니다.

    1. **`selectedCartItems`**

       * 장점: 주문 금액을 계산할 때 전체 `cartItems`를 한 바퀴 돌지 않고, Context로 관리된 `selectedCartItems`에 포함되었는지 여부만 확인하면 됩니다.
       * 단점: 선택되지 않은 `cartItems`는 전역적으로 관리되지 않습니다.
    2. **`cartItems`에 `isSelected` 타입을 추가한 방식**

       * 장점: 서비스를 더 넓게 생각할 때, 선택된 상품뿐만 아니라 선택되지 않은 상품들도 전역적으로 여러 곳에서 사용할 수 있습니다.
       * 단점: 리렌더링 시마다 백엔드에서 받아온 전체 `cartItems`에서 `isSelected`를 확인해야 하므로 주문 금액 계산 시 배열 전체를 한 바퀴 돌아야 합니다.
  * 최종적으로 저희는 다음 이유로 **1번 방식을 선택했습니다**.

    * 이번 미션은 상품 목록 페이지 없이 오로지 '장바구니' 페이지만 다루기 때문에 선택되지 않은 `cartItems`는 전역적으로 관리할 일이 없다고 판단했습니다.
    * 지금 미션의 요구사항이 장바구니에서 선택된 상품만을 계산하고, 페이지 이동 시에도 이 계산 금액만 필요하므로 전체 장바구니 상품이 아닌 `selectedCartItems`만 전역으로 관리하면 충분하다고 생각했습니다.
  * 하지만 실제 구현을 하다 보니 선택되지 않은 `cartItems`가 필요한 순간이 있었습니다. 이로 인해 상태 끌어올리기를 통해 props drilling이 발생했습니다. 예를 들어, 기존에는 `CartList`에서만 `cartItems` 상태가 필요했는데, 구현 과정에서 `CartHeader`에도 상품 수량 정보가 필요해졌습니다. 결국 공통 부모 컴포넌트인 `App`으로 상태를 끌어올리게 되었으며, 이 과정에서 `selectedCartItems` 대신 처음부터 `cartItems`로 Context를 설정할걸 그랬나 하는 생각이 들었습니다.


### 2) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점
* **테스트 목킹 - MSW vs mockImplementation**

  * 실제 구현한 상품 삭제, 상품 선택, 수량 변경 기능을 테스트하기 위해 서버 데이터가 있어야 하는지, 그리고 어떻게 테스트를 작성해야 하는지 고민했습니다.
  * 처음에는 실제 서버에 API 요청을 직접 보내어 이를 기반으로 테스트하면 되지 않을까 생각했지만, 그렇게 하면 네트워크 과정에서 발생할 수 있는 불안정성 문제, API 요청으로 인해 테스트 실행 속도가 느려질 수 있는 문제, 그리고 불필요한 비용이 발생할 수 있는 문제가 있을 것 같았습니다. 그래서 mock data를 생성하여 테스트를 진행하는 방식을 선택했습니다.
  * MSW를 사용하면 handler를 통해 간편하게 API 요청을 목킹할 수 있어서 MSW를 사용했습니다. 그러나 MSW를 사용하면서도 만약 실제 서버의 API 구조가 변경된다면, 테스트 코드를 변경하지 않아도 MSW 서버는 기존의 handler 기준으로 동작하여 테스트가 통과될 수도 있다는 의문이 들었습니다. 더불어, 테스트 작성을 위해 MSW 서버를 구축하는 게 과연 적절한가에 대한 고민도 있었습니다. 다른 크루에게 물어보니 mockImplementation이라는 방식도 있다고 했으나, 이미 MSW로 진행 중이어서 그대로 진행했습니다.

* **테스트 작성 과정에서의 어려움**

  * 테스트 작성에 아직 익숙하지 않고, 테스트 메서드도 잘 몰라 작성이 조금 번거로웠습니다. 예를 들어 장바구니에서 상품을 선택했을 때 주문 금액이 증가해야 하는데, 이를 테스트 코드로 어떻게 검증할 수 있을지 방법을 도출하는 과정에서 어려움을 겪었습니다. 일단 페어와 함께 열심히 작성했지만, 결국 시간상 주요 기능들마저도 전부 작성하지 못한 아쉬움이 남았습니다.

* **FSD 아키텍처**

  * 항상 고전적인 폴더 구조만 사용하다가, 이번에 페어의 제안으로 FSD 아키텍처를 적용하게 됐습니다. 제게는 너무 낯선 폴더 구조였지만, 페어가 이 아키텍처의 장점에 대해 기능별로 폴더를 나누는 방식이라고 설명해 주어서 흥미로웠습니다. 새로운 시도를 통해 다양한 경험을 얻고자 페어의 제안을 받아들이게 됐습니다.
  이번에 경험해 본 결과, 사실 이 정도 규모의 미션에서는 고전적인 아키텍처가 더 깔끔할 수 있겠다는 생각도 들었습니다. 다만, 향후 더 큰 규모의 프로젝트를 진행하게 된다면, FSD 구조가 더 구체적이고 체계적이라 원하는 파일을 찾기 훨씬 수월할 것 같다는 점을 느꼈습니다.


### 3) 이번 리뷰를 통해 논의하고 싶은 부분
앞서 2)에서 고민한 내용들과 연관지어, 효율적인 상태 관리 및 구조(Context API), 테스트 코드, 렌더링 최적화 등을 중심으로 리뷰를 받고 싶습니다.

또한, 현재 테스트 코드를 작성하는 방식(MSW를 사용하는 방법)이 적절한지, 방향성을 잘 잡고 있는지에 대한 의견을 듣고 싶습니다. 더불어 현업에서는 보통 테스트 코드를 어떤 방식으로 작성하는지도 궁금합니다.
(현재 코드에서는 MSW를 활용하여 Vitest + RTL로 mock 데이터를 기반으로 테스트를 진행했습니다. 현업에서도 Vitest + RTL 기반으로 테스트할 때 MSW를 사용하는지 궁금합니다!)

---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도뿐만 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. 클라이언트 상태관리

- 원본상태/파생상태를 적절히 구분하여 선언하였나요?
- React state 를 불필요하게 선언한 부분은 없나요?
- 상태 관리 로직의 책임이 적절히 응집/분리되었나요? (ex. reducer, hook, Context)

### 2. MSW/Test

- 주요 기능을 적절히 정의하였나요?
- 주요 기능/예외에 대한 테스트가 충분히 이루어졌나요?

woowa-euijinkk committed:
chore: 의존성 제거

eunwoo-levi committed:
chore: Emotion 라이브러리 설치

Co-authored-by: 이상희 <sang6299@naver.com>

eunwoo-levi committed:
docs: README 기능 요구 명세서 작성

eunwoo-levi committed:
feat: Navbar 컴포넌트 구현

eunwoo-levi committed:
styles: 전역 css 초기화

Co-authored-by: 이상희 <sang6299@naver.com>

eunwoo-levi committed:
feat: Cart 메인 페이지에서 큰 레이아웃 구현

eunwoo-levi committed:
chore: emotion tsconfig 전역 설정

eunwoo-levi committed:
feat: Custom Button 컴포넌트 구현

eunwoo-levi committed:
feat: Footer 컴포넌트 구현 및 Button컴포넌트 적용

eunwoo-levi committed:
feat: App 컴포넌트에 Footer 및 AppContent 적용

eunwoo-levi committed:
feat: CartHeader 컴포넌트 구현 및 적용

Co-authored-by: 이상희 <sang6299@naver.com>

eunwoo-levi committed:
feat: Cart feature endpoint 추가

eunwoo-levi committed:
chore: 파일명 변경

Co-authored-by: 이상희 <sang6299@naver.com>

eunwoo-levi committed:
feat: CartItemCard 컴포넌트 구현

eunwoo-levi committed:
feat: CartItemQuantitySelector 컴포넌트 구현

eunwoo-levi committed:
feat: CartList 컴포넌트 구현

eunwoo-levi committed:
styles: Button 컴포넌트 style 구현

eunwoo-levi committed:
feat: SelectInput 컴포넌트 구현

eunwoo-levi committed:
feat: CartList컴포넌트를 App에 적용 및 endpoint 추가

eunwoo-levi committed:
feat: 가격 정보 OrderPriceSummary 컴포넌트 구현

Co-authored-by: 이상희 <sang6299@naver.com>

eunwoo-levi committed:
feat: Shared 디렉토리에 Cart타입 정의

Co-authored-by: 이상희 <sang6299@naver.com>

eunwoo-levi committed:
feat: SelectedCart 정보를 가진 context 구현

eunwoo-levi committed:
feat: httpClient 유틸 함수 구현

eunwoo-levi committed:
feat: 백엔드로부터 CartItem 들고오는 GET API 로직 구현

eunwoo-levi committed:
feat: context를 통하여 CartItemCard의 input box를 통한 가격 계산 로직 구현

eunwoo-levi committed:
feat: 삭제 버튼 눌렀을 때 cartItem에 대해 DELETE API 호출 및 context에서 제거 로직 구현

eunwoo-levi committed:
feat: - 또는 +버튼을 눌렀을 때 CartItem 수량 업데이트 로직 구현

- 백엔드 API Patch 로직
- context 로직

Co-authored-by: 이상희 <sang6299@naver.com>

eunwoo-levi committed:
docs: 기능 요구 명세서 업데이트

eunwoo-levi committed:
feat: 장바구니 상품 없을 시 UI 보여주는 컴포넌트 및 로직 구현

eunwoo-levi committed:
refactor: 상품 없을 시 헤더 수정

eunwoo-levi committed:
feat: createBrowserRouter 사용하여 routes 구현

eunwoo-levi committed:
feat: 모든 페이지를 위한 Layout 컴포넌트 구현

eunwoo-levi committed:
feat: Error Page 구현

eunwoo-levi committed:
chore: CartPageFooter 파일 경로 이동

eunwoo-levi committed:
refactor: Navbar 컴포넌트에 handleClick 매개변수 전달

eunwoo-levi committed:
feat: main.tsx에 RouterProvider 적용

eunwoo-levi committed:
feat: ConfirmationPage 컴포넌트 적용

eunwoo-levi committed:
feat: 경로 URL 상수화 정의

eunwoo-levi committed:
chore: vitest 설치 및 세팅

Co-authored-by: 이상희 <sang6299@naver.com>

woowa-euijinkk committed:
docs: PR template

eunwoo-levi committed:
fix: vitest setup 에러 해결

Co-authored-by: 이상희 <sang6299@naver.com>

eunwoo-levi committed:
chore: msw 설치 및 환경 설정

eunwoo-levi committed:
chore: msw 설정 수정 및 mock data 추가

Co-authored-by: 이상희 <sang6299@naver.com>

eunwoo-levi committed:
refactor: data-testid 추가

eunwoo-levi committed:
test: 장바구니 테스트코드 추가

eunwoo-levi committed:
docs: README 기능 요구 명세서 업데이트

eunwoo-levi committed:
fix: router에 의해 base URL 설정 수정

Creative-Lee commented on __test__/cart.test.tsx:
```diff
@@ -0,0 +1,108 @@
+  beforeEach(() => {
+    renderApp();
+  });
+  it('장바구니에 상품이 하나도 없을 때 EmptyCartItemUI가 잘 보이는지', async () => {
```
테스트 명의 결을 다른 테스트들과 비슷하게 살짝 개선해볼까요?

sanghee01 replied:
시간상 반영하지 못했네요, step2때 개선해보겠습니다!!

Creative-Lee commented on package.json:
```diff
@@ -4,22 +4,27 @@
+    "deploy": "vite build --base=/react-shopping-cart/ --minify=false && npx gh-pages -d dist"
   },
   "dependencies": {
+    "@emotion/react": "^11.14.0",
+    "@emotion/styled": "^11.14.0",
```
[이미 여러번 들었을수도 있는 질문]

이모션 선택의 이유가 궁금합니다 ~!

sanghee01 replied:
미션에서 늘 그랬듯이 CSS Module, styled-components, emotion 중 한 가지를 선택해야했는데요,

CSS Module은 동적 스타일링이 번거롭고 타입 추론이 불가한다는 단점이 있어서 Styled-components와 emotion을 두고 고민을 했습니다.
두 개 다 사용하는 방식에서는 비슷하지만, emotion이 번들 사이즈가 작고, 런타임 성능도 더 좋다고 들어서 emotion으로 선택했습니다!

Creative-Lee commented on package.json:
```diff
@@ -32,9 +37,16 @@
+    "vite": "^5.2.0",
+    "vitest": "^3.1.4"
+  },
+  "msw": {
+    "workerDirectory": [
+      "public"
+    ]
```
👍

Creative-Lee commented on src/app/layout/Layout.tsx:
```diff
@@ -0,0 +1,10 @@
+import { Outlet } from 'react-router';
```
[궁금]
Outlet 활용 경험은 어땠나요~?

sanghee01 replied:
`Outlet`을 활용하게 된 이유는 장바구니 페이지와 결제 페이지에서 공통 스타일을 적용하고 싶었는데, 각 페이지마다 따로 CSS를 작성하기엔 중복이 발생해서 고민이었습니다. 근데 `Layout`이라는 컴포넌트를 만들어 스타일을 입히고, 각 페이지들을 `Outlet`에 렌더링되도록 함으로써 모든 페이지의 레이아웃 스타일을 일관되게 유지할 수 있어서 정말 좋았습니다!

Creative-Lee commented on src/shared/ui/SelectInput.tsx:
```diff
@@ -0,0 +1,12 @@
+
+export default function SelectInput({ ...rest }) {
+  return <SelectInputBox {...rest} />;
+}
```
이 컴포넌트는 어떤 목적의 컴포넌트인가요~? 
checkbox로 활용할거라면 type='checkbox' 라도 여기에 붙어야하지 않을까요~?

sanghee01 replied:
이 컴포넌트는 장바구니 체크박스 UI를 공통적으로 재사용하기 위해 SelectInput 컴포넌트로 만든 것입니다!

기존에는 사용처에서 `type='checkbox'`를 props로 넘기고 있었는데요, 다음과 같은 이유로 도밥 말씀대로 컴포넌트 내부에 작성하는 걸로 수정했습니다!

1. SelectInput 사용처에서 매번 type을 지정해야해서 번거로움
2. SelectInput 사용처에서 type 누락이 발생하면 예상치 못한 동작이 발생할 수 있음
3. SelectInput이라는 네이밍 자체가 checkbox 용도를 암시

또한 타입 정의도 추가했습니다

[refactor: SelectInput 컴포넌트 사용성 개선](https://github.com/woowacourse/react-shopping-cart/pull/357/commits/8f109e554426e62de0fc3f360e2a849de52154cd)

Creative-Lee commented on src/shared/type/cart.ts:
cart라는 도메인의 타입은 shared 레이어에 둘 코드가 아니라고 생각하는데요~!
예를들어 shared - ui - button의 결과는 조금 다른 결 같지 않나요?

Creative-Lee replied:
위쪽에 CartContext도 비슷한 의견입니다.

sanghee01 replied:
엇 그러네요..?! 페어할때 잘못 작성했나봅니다. 이부분은 늦게 확인해서 step2에 적용할게요!

Q. 
이것과 별개로 폴더 구조에 관해 궁금한게 있는데요, 현재 features/cart/ui 디렉토리를 보면 파일이 너무 많이 있어서 개인적으로 컴포넌트별로 폴더를 또 만들고 싶었는데요(컴포넌트와 해당 css파일 묶는 용도), 
페어는 그러면 너무 폴더층이 깊어지는 것 같다는 의견으로 적용하진 않았습니다. 
이에 대해 도밥은 어떻게 생각하시나요?

Creative-Lee replied:
정답은 없겠지만, 계층이 깊어지는것이 어떤것 때문에 문제가 되는지 물어볼 것 같아요 ㅎㅎ
워낙 IDE 선에서 파일 검색도 잘되기 때문에(일일히 디렉터리를 찍어서 찾는 경우가 거의 없는것 같아요) 불편함이 없다고 판단해서
저라면 말씀해주신 방식처럼 css 파일도 하나로 묶기위해 디렉터리를 하나 더 만들것 같습니다!

Creative-Lee commented on src/mocks/handler.ts:
```diff
@@ -0,0 +1,30 @@
+export const handlers = [
+  http.get(`${BASE_URL}/cart-items*`, () => {
+    return HttpResponse.json({ content: cartItems });
+  }),
```
개별 핸들러도 함수로 분리하여 네이밍해주면 가독성이 더 좋아질 것 같아요.

sanghee01 replied:
오호 확실히 분리하는게 가독성 면에서 나을 것 같아서 분리했습니다!

[refactor: msw handler 개별 함수로 분리](https://github.com/woowacourse/react-shopping-cart/pull/357/commits/035939d6462d443aaa61f0703c92b8a5f4ab2d69)

Q. 
`updateCartItemQuantity`에서 매개변수 중 `request` 타입을 어떻게 해야할 지 모르겠습니다. 

ai를 통해 해결하려고 했는데, 어떻게 지정하든 타입이 올바르지 않다는 에러가 발생하는데요, 
MSW에서는 타입을 명시적으로 정의하지 않고 자동 추론에 맡기는 것이 가장 좋은 방법이라고 계속해서 말하길래 결국 타입을 지정하지 않았습니다. 
이게 맞는 말인지, 도밥이라면 어떻게 하실건지 궁금합니다.

Creative-Lee replied:
지금은 타입에러가 발생하는 상태인데,

```ts

export const mockPatchCartItem = http.patch(`${BASE_URL}/cart-items/:id`, async ({ params, request }) => {
  const id = Number(params.id);
  const body = await request.json();
  const { quantity } = body as { quantity: number };

  const itemIndex = cartItems.findIndex((item) => item.id === id);
  if (itemIndex !== -1) {
    cartItems[itemIndex].quantity = quantity;
  }

  return HttpResponse.json({ message: '장바구니 아이템이 변경되었습니다.' });
});

```
위 코드 처럼 http.patch 호출을 포함하는 핸들러 자체로 분리하면 말씀해주신 타입적 에러가 발생하지는 않는 것 같습니다!
(코멘트의 의도도 위와 같습니다. - 제가 함수로 분리하라고 해서 혼동을 드린 것 같네요..! )

sanghee01 replied:
아핫 그렇군요..! 이렇게하면 사용처(browser.ts, server.ts)에서 어떤 핸들러를 사용하는지 명시적으로 확인 가능하겠네요!

step2에서 적용해보겠습니다

Creative-Lee commented on src/pages/CartPage/App.tsx:
CartPage이면 CartPage여야할 것 같은데, App이네요
이유가 있나요?

sanghee01 replied:
앗 해당 컴포넌트가 기존엔 `루트` 경로에 있었다가 추후 `Router`를 도입하면서 `CartPage` 폴더로 이동시켰는데요, 그 과정에서 컴포넌트명을 변경하는걸 깜빡 잊었네요!

[refactor: App -> CartPage 컴포넌트명 변경](https://github.com/woowacourse/react-shopping-cart/pull/357/commits/9b61c974c2d6f0efce32279f0302a126c63eb3ac)

Q.
근데 사소하게 궁금한게 있는데요! 
이전 미션들에서는 `App` 컴포넌트가 최상단 부모 역할을 했는데, 지금은 `Router` 도입으로 페이지 개념이 돼서 `App` 컴포넌트의 존재? 필요성이 필요 없어진 느낌이네요. 

즉, 이제 `App` 네이밍의 컴포넌트는 없는건데, 이게 자연스러운 현상인걸까요? 뭔가 `App` 은 항상 있었던 존재인데 이제 없어서 ‘이래도 되는건가?’라는 생각도 들어서요!

Creative-Lee replied:
자연스럽게 느껴집니다. 
router.tsx로 인해 최상단 컴포넌트 개념이 사라졌고, 앱내 공통 처리는 main.tsx(혹은 index.tsx일때도 있는)에서 하게되겠죠.

만약 엄밀하게 구분하여 앱내 공통 처리(e.g. `<React.StrictMode/>`, `<QueryProvider/>` 등)의 개념과 
도메인 개념(e.g. `<SelectedCartItemsProvider />` 등) 을 구분하고 싶다면 router에 최상위 페이지 컴포넌트 개념의 App이 다시 등장할수도 있겠고요~! 
정답은 없습니다ㅎㅎ

Creative-Lee commented on src/pages/CartPage/App.tsx:
```diff
@@ -0,0 +1,58 @@
+      }
+    };
+    fetchCartItems();
+  }, [selectedCartItems]);
```
컴포넌트에 여러 로직들을 분리하는 연습을 해보면 좋을 것 같은데요~!

get api 관련 상태와 이를 set 하기위한 effect를 useCartItems등의 hook으로 묶어주는것부터 시작하면 좋겠습니다.

Creative-Lee replied:
위의 코멘트를 반영했다면, 
이어서 지난 미션에서 설계했던 데이터 패칭 hook을 적용해봐도 좋을 것 같아요.
(만약 아직 지난 미션도 함께 진행하고 있고, 완성되지 않았다면 위 코멘트만 반영해도 좋아요.)

sanghee01 replied:
말씀대로 관심사 분리가 잘 안되어있었네요 🥲

해당부분은 CartItemsProvider에서 컴포넌트 마운트시 fetchCartItems가 실행되도록 하는걸로 수정했습니다!

데이터 패칭 hook은 step2에 시간이 괜찮다면 확인해보겠습니다! 

[refactor: 전반적인 상태 관리 구조 변경](https://github.com/woowacourse/react-shopping-cart/pull/357/commits/f3ecb00f8c6fd7501f672b6aa019eb21add3803c)

Creative-Lee commented on src/main.tsx:
```diff
@@ -1,10 +1,31 @@
+async function enableMocking() {
+  // if (process.env.NODE_ENV === 'development') {
+  //   return;
+  // }
```
~🗑️

sanghee01 replied:
Q. 
이는 개발할 때 Mocking이 안된 환경을 보고싶을 때를 대비해서 일부러 주석으로 넣었던 코드인데요, 이런 경우에도 지우고 리뷰요청을 보내는게 맞을까요?! 즉, 설명을 위한 주석 제외하고 보통 모든 주석은 지우고 리뷰요청 보내는건지 궁금해요! 

리뷰 요청 당시에도 이걸 지울까 말까 고민했었는데, 일단 지웠습니당!
[refactor: 주석 제거](https://github.com/woowacourse/react-shopping-cart/pull/357/commits/b0b81756ba2d5a4995e20b703c3be8f527ac5831)

Creative-Lee replied:
아하 그런 의도라면 상단에 주석을 추가로 달아주셔도 좋겠습니다~!
리뷰 요청에 룰이라기보다는, 동료가 이 코드의 의도를 알 수 있냐 없냐를 기준으로 행동해주시면 될것 같아요~!

(별개로 `개발할 때 Mocking이 안된 환경을 보고싶을 때` 는 매번 주석을 풀었다 걸었다 하는 것 보다는,
환경변수를 활용하여 `npm run dev:msw` 등의 스크립트를 추가해 해결해볼 수 있을 것 같습니다.)

Creative-Lee commented on src/app/routes/routes.tsx:
```diff
@@ -0,0 +1,16 @@
+export const router = createBrowserRouter(
+  createRoutesFromElements(
+    <Route path={ROUTES.ROOT} element={<Layout />}>
+      <Route path={ROUTES.CART} element={<App />} />
```
지금 이대로면 중첩되는 라우트 경로를 가지게 되는것 같은데요~! 맞나요? 
관련해서 `index` 라는 prop이 도움이 될수도 있겠습니다.

sanghee01 replied:
오 안그래도 라우트 경로가 중첩돼서 뭔가 찜찜했는데 `index` prop 라는게 있군요!   
이게 왜 가능한지 궁금해서 찾아봤더니, `index routes`는 부모 라우트의 `<Outlet/>`에서 부모와 정확히 동일한 URL에 렌더링되는 방식이군요! 

[refactor: 중복된 route path index prop 사용](https://github.com/woowacourse/react-shopping-cart/pull/357/commits/161fe5017236061d3cec3bfd180e81f4221e2e48)

Creative-Lee commented on src/app/routes/routes.tsx:
```diff
@@ -0,0 +1,16 @@
+import ConfirmationPage from '../../pages/ConfirmationPage/ConfirmationPage';
+
+export const router = createBrowserRouter(
+  createRoutesFromElements(
```
<img width="661" alt="image" src="https://github.com/user-attachments/assets/37335373-43c6-4ffe-8f3a-dfc7336b6b6c" />

createBrowserRouter가 요구하는 타입으로 작성하지 않고,
createRoutesFromElements를 사용하면서 까지 <Route>로 이루어진 형태를 유지한 이유가 있을까요?

sanghee01 replied:
저도 페이먼츠 미션에서 react-router 공식문서 참고하면서 적용했을땐, 다음 코드처럼 배열 안에 객체를 넣는 방식으로 작성했는데요,  (지금보니 이때 이미 index 기능을 써봤었네용.. ㅎ.ㅎ)
```tsx
const router = createBrowserRouter([
  {
    path: ROUTE_PATH.ROOT,
    element: <RootLayout />,
    children: [
      {
        index: true,
        element: <CardRegistrationPage />
      },
      {
        path: ROUTE_PATH.CARD_REGISTRATION_COMPLETED,
        element: <CardRegistrationCompletedPage />
      }
    ]
  }
]);
```

이번에 페어 프로그래밍을 진행하면서, 페어가  `createRoutesFromElements` 를 사용한 JSX 기반의 선언적 라우트 구성이 더 직관적이고, 라우팅 구조를 관리하거나 유지보수할 때 더 효율적이라고 주장했는데, 이에 설득당해서 이렇게 진행했습니다. 

Q.
도밥은 이에대해 어떻게 생각하시나요? 그래도 createBrowserRouter가 요구하는 타입으로 작성하는게 낫다고 생각하시나요?

Creative-Lee replied:
납득가능한 이유라고 생각합니다 ㅎㅎ
그런데 살짝 궁금한것은(진짜 궁금), 지금의 방식(`createBrowserRouter`, `createRoutesFromElements` 둘다 쓴 방식)
아래 방식으로 명식하는 거랑 하고 어떤 차이가 있는걸까용? ㅋㅋ

```tsx
    <BrowserRouter> // 그냥 이걸 쓴거랑 어떤차이가?
      <Route path={ROUTES.ROOT} element={<Layout />}>
        <Route index element={<CartPage />} />
        <Route path={ROUTES.ORDER_CONFIRMATION} element={<OrderConfirmationPage />} />
        <Route path={ROUTES.ORDER_SUCCESS} element={<OrderSuccessPage />} />
        <Route path="*" element={<NotFoundPage />} />
      </Route>
    </BrowserRouter>
```

sanghee01 replied:
저도 좀 더 자세히 알고 싶어서 찾아봤는데요, 차이가 있습니다!


### 1. `createBrowserRouter` 와 `createRoutesFromElements` 

우선 `BrowserRouter` 와의 차이점을 말씀드리기 앞서, 도밥이 앞서 다음과 같이 말씀 주셨는데요, 

> createBrowserRouter가 요구하는 타입으로 작성하지 않고, createRoutesFromElements를 사용하면서 까지 로 이루어진 형태를 유지한 이유가 있을까요?
> 

`createBrowserRouter`는 RouteObject[] 타입을 인자로 받는데요,

`createRoutesFromElements`는 JSX로 작성한 <Route> 트리를 RouteObject[]로 변환해주는 함수여서, 결과적으로 `createBrowserRouter`가 요구하는 타입을 맞추는 셈입니다!

즉, `createRoutesFromElements` 는 내부적으로는 결국 RouteObject[]로 변환되므로, `createBrowserRouter` 와는 기능상 차이가 없습니다!


<br/>

### 2. `BrowserRouter` 와 `createRoutesFromElements` 의 차이점

먼저, `createRoutesFromElements` 는 내부적으로는 결국 RouteObject[]로 변환되지만,  `BrowserRouter` 는 라우터 객체가 없다는 차이점이 있습니다.

또, `createRoutesFromElements` 는 `BrowserRouter` 과 달리 다음과 같이 loader, action, errorElement 등 Data API 를 사용할 수 있습니다!

```jsx

const router = createBrowserRouter(
  createRoutesFromElements(
    <Route path="/" element={<Layout />}>
      <Route
        index
        element={<CartPage />}
        loader={cartLoader} // Data API 사용
        action={orderConfirmationAction} 
        errorElement={<OrderError />}
      />
	    ...
    </Route>
  )
);
```

### 관련 링크 
참고하면 좋을 링크 공유드립니다!

[React-Router 공식문서: tutorial, jsx-routes부분](https://reactrouter.com/6.30.1/start/tutorial#jsx-routes)
[React-Router 공식문서: Function createRoutesFromElements](https://api.reactrouter.com/v7/functions/react_router.createRoutesFromElements.html)
[React Router: Routes, Loaders and Errors](https://dev.to/vikram-boominathan/react-router-routes-loaders-and-errors-1nee)

Creative-Lee commented on src/features/cart/ui/CartItemCard.tsx:
```diff
@@ -0,0 +1,92 @@
+import { useSelectedCartContext } from '../../../shared/context/useCartContext';
+import { deleteCartItem } from '../api/deleteCartItem';
+
+const deleteButtonCSS = css`
+  width: 40px;
+  height: 28px;
+  background-color: #fff;
+  border: 1px solid rgba(0, 0, 0, 0.1);
+  border-radius: 4px;
+  font-size: 13px;
+  font-weight: 500;
+  transition: background-color 0.2s ease, border-color 0.2s ease;
+  &:hover {
+    background-color: #f8f8f8;
+    border-color: rgba(0, 0, 0, 0.2);
+  }
+`;
```
DeleteButton은 컴포넌트로 분리하지 않았는데, 이유가 있나요?

sanghee01 replied:
deleteButton은 컴포넌트라기 보단 css 코드인데요, emotion pragma라고 css을 props로 전달하기 위해 사용됩니다! 
공용 컴포넌트인 Button을 사용하면서도 추가적으로 css를 입히고 싶어서 이렇게 사용했습니다!

```tsx
  <Button title="삭제" css={deleteButtonCSS} onClick={handleCartItemDelete} />
```

Creative-Lee commented on src/shared/context/useCartContext.tsx:
1. 파일명이 훅 이름과 조금 다르군요
2. 응집을 위해 `SelectedCartProvider` 하단에 같이 작성해도 좋을 것 같습니다.(파일 하나로)

sanghee01 replied:
파일명을 올바르게 변경하는걸로 수정했습니다! 
`useCartContext` → `useSelectedCartContext` 

[refactor: useSelectedCartContext 파일명 컴포넌트 명과 맞도록 수정](https://github.com/woowacourse/react-shopping-cart/pull/357/commits/81de19b7171676104c2220ea746dbe32569c3fe6)

Q.
한 파일로 응집시키는것도 고민을 해봤는데요, 
`SelectedCartProvider`는 상태 관리와 비즈니스 로직 구현, `useSelectedCartContext`는 안전한 접근 인터페이스 제공이라는 서로 다른 추상화 레벨의 책임을 담당해서, 분리하는게 단일 책임 원칙 입장에서 맞는 것 같다고 판단해서 응집시키진 않았습니다! 

도밥은 응집을 위해 한 파일로 묶는 의견을 주셨는데, 왜 그렇게 생각하셨는지 궁금해요!! 제가 놓친 부분이 있을까봐요!

Creative-Lee replied:
컨텍스트 + 프로바이더 와 `해당 컨텍스트`를 랩핑하는 hook이기에 한 파일에 작성해도 괜찮다는 의견이었습니다. 
코드 관리 측면에서 모아놓는게 유지보수가 쉽겠다~ 라는 접근이었어요. (컨텍스트가 더 많아진다면?)

추상화 레벨이라는 단어까지 필요한지는 잘 모르겠습니다만
책임을 분리하기 위함이라는 목적이 분명하니, 이대로 두어도 좋습니다 ㅎㅎ

Creative-Lee commented on src/pages/CartPage/App.tsx:
```diff
@@ -0,0 +1,58 @@
+      }
+    };
+    fetchCartItems();
+  }, [selectedCartItems]);
```
useEffect의 의존배열을 조건문처럼 사용하지 않았으면 좋겠습니다. 
(상품의 선택여부 변경과 CartItem fetch가 관련이 있었나요?)

sanghee01 replied:
마찬가지로 PR 본문에 자세히 작성했습니다! 
리렌더링을 트리거 할 수 있는 엉뚱한 변수를 끼워 넣었던 것 같네요. cartItems를 refech하는 기능을 만들어서 해결했습니다!

Creative-Lee commented on src/features/cart/ui/CartItemQuantitySelector.tsx:
```diff
@@ -0,0 +1,85 @@
+    };
+
+    handleCartItemQuantityUpdate();
+  }, [cartQuantity]);
```
여기도 의존하는 값은 훨씬 많은데, 수량 변경시에만 조건부로 호출하기위해 의존배열을 사용하고 있습니다. 

공식문서를 한번 읽고 가도 좋겠습니다.
https://ko.react.dev/reference/react/useEffect#specifying-reactive-dependencies

Creative-Lee commented on src/features/cart/ui/CartItemQuantitySelector.tsx:
```diff
@@ -0,0 +1,85 @@
+          alert('장바구니 아이템 수량 업데이트에 실패했습니다. 다시 시도해주세요.');
+        }
+      }
+    };
```
관리하는 상태들이 다소 얽혀있어서 흐름이 어렵다고 생각이 듭니다.

effect의 각 함수는 아래와 같은 역할을 하는것 같은데요
- 클라이언트 상태 업데이트
- Context로 관리하는 상태 업데이트
- 서버api 호출

이 부분에 대해서 상추의 상태 관리 설계의도가 궁금합니다. 
수량을 변경하는 과정에서는 어떤 상태를 변경해야하며, 어떤 데이터의 흐름으로 UI를 업데이트 하려고 했나요?

sanghee01 replied:
해당 부분은 수정했는데요, PR 본문에 이에 대해 고민한 과정을 작성했습니다!

사실 페어할 당시에는 서버 상태, 클라이언트 상태와 같은 개념이 잡혀있지 않았습니다. 그저 잘 동작하도록 짜는거에만 집중했던 것 같습니다.

도밥의 말씀과, 한주동안 고민했던 내용을 바탕으로 다시 코드를 보니 설계가 많이 잘못됐음을 느낄 수 있었는데요(뭔가 상태 관리나 리렌더링 호출 타이밍이 꼬여있음),
 
이번에 이 사태를 해결하기 위해 상태 흐름, 컴포넌트 리렌더링 순서 및 트리거, 역할과 책임에 대해 많이 고민했던것 같네요.

Creative-Lee commented on src/features/cart/ui/CartItemQuantitySelector.tsx:
```diff
@@ -0,0 +1,85 @@
+export default function CartItemQuantitySelector({
+  isSelected,
+  updateSelectedCartItem,
+  cartItem,
+  quantity,
```
cartItem 안에 수량이 있는데, 별개의 prop으로 받을 필요가 없을 것 같습니다.

sanghee01 replied:
헉스 그렇네요!!

[refactor: 불필요한 매개변수 제거](https://github.com/woowacourse/react-shopping-cart/pull/357/commits/1c6c21d450ba04e1629e17fc825090344097a524)

Creative-Lee commented on src/features/cart/ui/CartList.tsx:
```diff
@@ -0,0 +1,35 @@
+
+interface CartListProps {
+  cartItems: CartItem[];
+  setCartItems: React.Dispatch<React.SetStateAction<CartItem[]>>;
```
setState를 직접 전달하는 것은 자식에게 부모의 내부 구현을 알려주는 것과 같다고 생각합니다. 더불어 상태변경의 흐름을 추적하기도 어려워집니다. 
부모에게서 상태를 업데이트 할 수 있는 랩핑된 함수를 전달받는 개념으로 변경해보면 좋겠습니다.

sanghee01 replied:
그렇네요! 추상적인 인터페이스가 아닌, 구체적인 구현 방식에 의존하고 있다는 DIP 위반 문제가 있었네요!
또한 외부의 상태를 직접 제어해서 캡슐화를 위반한다는 문제가 있었네요.

CartItemsProvider에 increaseCartItemQuantity, decreaseCartItemQuantity라는 함수를 만들어서 이를 통해 상태를 업데이트 할 수 있도록 했습니다!

[refactor: 전반적인 상태 관리 구조 변경](https://github.com/woowacourse/react-shopping-cart/pull/357/commits/f3ecb00f8c6fd7501f672b6aa019eb21add3803c)

Creative-Lee commented on src/features/cart/ui/CartList.tsx:
```diff
@@ -0,0 +1,35 @@
+  return (
+    <S.CartListContainer>
+      <S.AllSelectContainer>
+        <SelectInput type='checkbox' onChange={handleAllCartItemsInSelected} />
```
현재 전체 선택 버튼에 버그가 존재합니다. 
- 개별 상품의 체크박스로 모두 선택한 경우에 전체선택 버튼은 체크되지 않음.
- 즉 개별 선택여부와 전체선택 버튼이 동기화 되지 않음.

전체 선택 버튼은 check prop등을 받지 않습니다. 즉, 상태로 컨트롤하는 컴포넌트가 아니기에 위와 같은 동기화 자체를 할 수 없습니다.

버그를 수정해보셔요

sanghee01 replied:
전체 선택 체크박스를 다음 코드와 같이 제어 컴포넌트로 변경해서 해결했습니다! 

기존에는 전체 선택 체크박스가 개별 선택 상태와 동기화되지 않는 비제어 컴포넌트 형태였어서, 개별 아이템들의 선택 상태가 변경되어도 전체 선택 버튼이 이를 감지하지 못했었네요!

```tsx
const isAllSelected = cartItems.length > 0 && SelectedCartItems.length === cartItems.length;

<SelectInput checked={isAllSelected} onChange={handleAllCartItemsInSelected} />

```

1. `checked` prop을 통해 선택 상태를 명시적으로 제어
2. `SelectedCartItems.length === cartItems.length` 조건으로 실시간 동기화
3. 개별 선택/해제 시 전체 선택 버튼 상태가 즉시 반영


[feat: 페이지 첫 진입시 모든 상품이 선택 돼있도록 & 개별 상품 클릭 다 되면 전체선택 버튼 활성화](https://github.com/woowacourse/react-shopping-cart/pull/357/commits/0b0b899f9970383e5f06379fd897ffc97cbcba05)

Creative-Lee commented on src/features/cart/ui/CartItemCard.tsx:
```diff
@@ -0,0 +1,92 @@
+    updateSelectedCartItem(cartItem, cartItem.quantity);
+  };
+
+  const isSelected = selectedCartItems.findIndex((item) => item.id === cartItem.id) === -1 ? false : true;
```
index를 사용할것이 아닌데, index로 찾는 이유가 있나요? 
find가 조금 더 명시적이지 않을까 싶습니다.

sanghee01 replied:
말씀대로 find를 사용하는게 더 명시적이어서 수정했어요! 페어 당시에는 생각하지 못했던 것 같아요 

[refactor: selected 여부 찾는 함수 명시적으로 수정](https://github.com/woowacourse/react-shopping-cart/pull/357/commits/84b5d93bddb6878b15564353749a0a03f00122f1)

Creative-Lee commented on src/features/cart/ui/CartItemCard.tsx:
```diff
@@ -0,0 +1,92 @@
+
+export default function CartItemCard({ cartItem, setCartItems }: CartItemCardProps) {
+  const { selectedCartItems, updateSelectedCartItem, removeSelectedCartItem } = useSelectedCartContext();
+  const handleSelectedCartItemUpdate = (e: React.ChangeEvent<HTMLInputElement>) => {
```
습관처럼 handle prefix를 붙이고 있지 않은지 생각해봅시다. 

`updateSelectedCartItem` 이라고 네이밍해도 충분하지 않나요?

sanghee01 replied:
이벤트 핸들러 함수는 관습적으로 `handle` 접두어를 붙이는걸로 알고 있습니다! [react 공식문서 - 이벤트 핸들러 추가하기](https://ko.react.dev/learn/responding-to-events)에서도 나와있는 부분이고요!

현재 `handleSelectedCartItemUpdate` 가 onChange 이벤트를 처리하는 핸들러 역할로 사용하고 있어서 이렇게 네이밍 했습니다. 

도밥은 왜 그렇게 생각하셨나요?! 혹시 제가 놓친 부분이 있을까요?!

Creative-Lee replied:
저도 네이밍에 정답은 없기에 팀내 컨벤션(혹은 관습)을 잘 따르면 된다고 생각하는 한편, 
컴포넌트 return문 상단에 선언하는 함수는 대게 핸들러 함수이기에 handle 이라는 prefix가 없이도 충분히 그 의미를 드러낼 수 있다고 생각합니다~!

비슷한 코멘트를 드렸을 때 해당 문서를 제시해주곤 하시는데요~!  문서를 토대로 관습을 따르기 위함이라면 좋다고 생각합니다ㅎㅎ
(이런 의미에서 문서에서도 이렇게 해라! 가 아니고 이런식으로 하곤한다~ 라는 늬양스로 말한게 아닐까 싶네요)

Creative-Lee commented on src/features/cart/ui/CartItemCard.tsx:
```diff
@@ -0,0 +1,92 @@
+  return (
+    <S.CartItemContainer data-testid='cart-item-card'>
+      <S.CartItemHeader>
+        <SelectInput type='checkbox' onChange={handleSelectedCartItemUpdate} checked={isSelected} />
```
페이지 로드 시 기본으로 모든 상품이 선택된 상태여야 합니다.
한번 반영해봅시다ㅎㅎ

<img width="459" alt="image" src="https://github.com/user-attachments/assets/f407daaa-b187-4c74-b104-4c045f0e5410" />

sanghee01 replied:
useEffect를 통해 페이지 첫 진입시 모든 cartItems가 selected 상태에 담기도록 했습니다!  

```tsx
useEffect(() => {
    setSelectedCartItems(cartItems);
  }, [cartItems]);
```

[feat: 페이지 첫 진입시 모든 상품이 선택 돼있도록 & 개별 상품 클릭 다 되면 전체선택 버튼 활성화](https://github.com/woowacourse/react-shopping-cart/pull/357/commits/0b0b899f9970383e5f06379fd897ffc97cbcba05)

다만 수량 변경시에도 cartItems의 quentity 변경으로인해 의존성이 트리거돼서 전체선택이 되는 버그가 발생하더라고요.
그래서 아래와 같이 초기 상태인지 확인하는 상태(init)도 추가했습니다! 

또한, 이렇게만 하면 페이지 진입시 전체 선택이 안되길래 원인을 파악했더니, 장바구니 데이터가 아직 로드가 안돼서 그런거였습니다. 따라서 `cartItems.length > 0`조건을 넣어 해결했습니다!

```tsx
const [init, setInit] = useState(false);

useEffect(() => {
    if (!init && cartItems.length > 0) {
      setSelectedCartItems(cartItems);
      setInit(true);
    }
  }, [cartItems, init]);
```

[fix: 페이지 진입시 상품목록 전체 선택 로직 수정](https://github.com/woowacourse/react-shopping-cart/pull/357/commits/a9db661d5d1691f9d5f3d051431a00975d308340)

Creative-Lee reviewed (changes requested):
안녕하세요 상추 ㅎㅎ 오랜만에 다시뵙네요 👋 
어느덧 마지막 미션인데 힘내봅시다~!

---

상태 설계 의도에 대해서 먼저 이야기 해보면,
> **Context API로 관리할 상태** - `selectedCartItems` vs. `cartItems`
> 각각 다음과 같은 장단점이 있어 고민하게 되었습니다.

단점에 대해서는 약간?의 의문은 있지만, 전체적으로 장단을 잘 정리해주신 것 같아요. 
(e.g. 전체 리스트를 순회하는것이나, 그렇지 않는 것이 실제로 어떤 성능차이 있을까? - 일반적으로 장바구니에 몇개나 담지? 등의 의문)

>최종적으로 저희는 다음 이유로 1번 방식을 선택했습니다.

적어주신 부분에 대해서는 최종 선택의 이유도 합리적이라고 생각합니다.
다만, 조금의 의문이 있어서 코멘트 남겼으니 확인해주세요 ㅎㅎ

> 하지만 실제 구현을 하다 보니 선택되지 않은 `cartItems`가 필요한 순간이 있었습니다. 

이런 상황에서는 api로 얻을 수 있는 장바구니 데이터를 전역으로 활용할 수 있게 구현 변경을 고려해 볼 수도 있겠습니다. 

(설계한대로 완벽하게 돌아갔으면 좋겠지만, 그렇지 못한경우가 훨씬 많은것 같아요 ㅎㅎ)

---

이어서, 고민하셨던 부분과 질문도 보면...

> FSD 아키텍처
> 항상 고전적인 폴더 구조만 사용하다가, 이번에 페어의 제안으로 FSD 아키텍처를 적용하게 됐습니다. 제게는 너무 낯선 폴더 구조였지만, 페어가 이 아키텍처의 장점에 대해 기능별로 폴더를 나누는 방식이라고 설명해 주어서 흥미로웠습니다. 새로운 시도를 통해 다양한 경험을 얻고자 페어의 제안을 받아들이게 됐습니다.
 이번에 경험해 본 결과, 사실 이 정도 규모의 미션에서는 고전적인 아키텍처가 더 깔끔할 수 있겠다는 생각도 들었습니다. 다만, 향후 더 큰 규모의 프로젝트를 진행하게 된다면, FSD 구조가 더 구체적이고 체계적이라 원하는 파일을 찾기 훨씬 수월할 것 같다는 점을 느꼈습니다.

새로운 시도는 언제나 좋습니다 👍 
시도를 했기에 디렉터리 구조에 대한 장단점도 파악하고 인사이트를 얻을 수 있었던게 아닐까요? 😄 
또한 상추의 의견에도 동의합니다. 지금정도의 구조에서는 크게 이점을 느끼기 어려울수도 있겠어요. 
정해진 FSD의 구조에 맞추기 전에, 단순하게 기능별로 디렉터리를 나누고 한곳에 모아놓는 행위자체를 시도해봐도 좋겠습니다.


> 테스트의 방향성 및 mocking 등

실제 서버 요청 대신 mocking으로 테스트를 하는 방법들 중, MSW를 활용해주신것도 납득 가능했습니다. 서버 mocking으로 활용하는 김에 테스트에도 활용해 볼 수 있다는 생각은 나쁘지 않은 것 같아요. 적절하게 선택해주셨다고 생각합니다. 

서버의 응답이 바뀌었을 때 여전히 통과될 수 있는 테스트와 관련해서는 조금 생각을 해보았는데, 
실제 변경된 응답 타입에 대한 수정전에 테스트가 성공하는 현상은 mockImplementation이나, msw 둘다 똑같이 발생하지 않을까? 싶네요.
더 생각을 이어가보면 
협의를 통해 응답 스펙이 바뀌었을 경우에는 클라이언트 개발자가 api response type을 수정하게 될것이고, mock handler도 이 api Response 타입을 따를테니 type에러가 발생할것 같습니다. 이 경우에는 수정을 해주면 될 것 같아요. 
이어서 만약 협의 없이 잠수함 패치를 한 경우라면 mockImplementation이나, msw 어떤것을 선택하던지 런타임 에러를 방지할 수는 없을 것 같습니다. (이때는 런타임 에러와 관련된 도구들을 활용해야할 것 같고요)


> 테스트 작성 과정에서의 어려움

많이 해보지 않았기에 익숙하지 않은 것이잖아요~! 너무 아쉬워 하진마셔요 😄 
익숙하지 않은 시기도 오래 가진 않을거에요. (얼마 뒤엔 아무생각없이 테스트를 작성하고 있을지도 모르죠ㅋㅋㅋ)

> 현업에서는...

vitest + RTL + MSW은 현업에서도 충분히 활용가능한 테스트 셋업이라고 생각합니다!
여러 선택지 중 하나라고 생각해요~ ㅎㅎ

---

고생많으셨습니다 상추!
마지막까지 힘내봅시다~ 💪

sanghee01 committed:
chore: 필요한 패키지 설치, 안쓰는 패키지는 제거

sanghee01 committed:
refactor: App -> CartPage 컴포넌트명  변경

sanghee01 committed:
refactor: msw handler 개별 함수로 분리

sanghee01 committed:
chore: prettier 설정

sanghee01 committed:
refactor: useSelectedCartContext 파일명 컴포넌트 명과 맞도록 수정

sanghee01 committed:
refactor: 주석 제거

개발 환경일때는 모킹을 안하도록 설정하는 주석 제거

sanghee01 committed:
refactor: 불필요한 매개변수 제거

quantity는 이미 cartItem 객체에 있는 정보

sanghee01 committed:
refactor: SelectInput 컴포넌트 사용성 개선

sanghee01 committed:
refactor: 앞서 지웠던 매개변수 안지운거 있어서 제거

sanghee01 committed:
refactor: selectedCartItemsContext 네이밍 좀 더 명시적으로 수정

sanghee01 committed:
refactor: 전반적인 상태 관리 구조 변경

- CartItems Context 생성(서버 상태 원천)
- SelectedItems Context 리팩토링(클라이언트 상태 원천)

sanghee01 committed:
refactor: 네이밍 수정

sanghee01 committed:
feat: 페이지 첫 진입시 모든 상품이 선택 돼있도록 & 개별 상품 클릭 다 되면 전체선택 버튼 활성화

sanghee01 committed:
fix: 페이지 진입시 상품목록 전체 선택 로직 수정

수량 변경시 전체선택되는 문제가 발생함. 해당 useEffect안에 cartItems가 의존성 배열 안에 있었는데, 수량 변경시 이 값이 변경되므로 전체선택된거임.
그래서 초기 진입 여부 상태를 만듦

sanghee01 committed:
refactor: 테스트코드 미흡한 부분 수정

- 계산 타이밍 이슈로 waitFor을 넣었음
- 배송비 판별을 정확히 탐지하기 위해 하드 코딩 가격으로 수정

sanghee01 committed:
refactor: selected 여부 찾는 함수 명시적으로 수정

sanghee01 committed:
refactor: 중복된 route path index prop 사용

sanghee01 commented:
안녕하세요 도밥! 
덕분에 많이 고민하고 학습할 수 있었던 시간이었습니다.

전반적으로 기존 상태 관리 설계가 잘못 된것 같아 구조를 좀 수정했습니다. 
좀 미흡한 부분이 있을 것 같은데, 혹시나 의문이 들거나 개선하면 좋을 사항 있으면 말씀주시면 감사하겠습니다!

시간상 일부 코멘트는 답장을 하지 못했는데, 시간날 때 다시 답장하겠습니다!

### 상태 관리 구조 변경 이유

기존의 상태 관리 방식에서는 `selectedItems` Context 내부에서 `cartItems`와 `selectedCartItems` 로직이 혼재되어 있었습니다. 이로 인해 두 데이터 간의 연관 관계가 불분명했고, 각 파일(컴포넌트 및 훅)의 역할과 책임 역시 명확하지 않았습니다. 

특히 `useEffect`에 반응형 상태가 아닌 외부 데이터를 넣어 강제로 리렌더링을 유발하는 방식으로 작성되었습니다. 페어 프로그래밍 당시에는 '반응형 값'에 대한 이해가 부족했기 때문에 [도밥의 코멘트](https://github.com/woowacourse/react-shopping-cart/pull/357#discussion_r2116089630)와 같은 문제가 발생했던 것 같습니다.

정리하자면, 기존 코드는 React의 원칙인 **단일 진실의 원천(SSOT)**, **파생 상태 활용**, **명시적인 상태 변경**, 그리고 **Effect의 올바른 활용** 원칙을 제대로 지키지 못하고 있었습니다. 이에 따라 구조를 명확히 하고자 전체적인 상태 관리 방식을 재구성했습니다.

### 변경 과정

먼저 기존에는 `cartItems`의 데이터를 다시 가져오는(refetch) 함수가 없어, [도밥이 지적하신 것처럼](https://github.com/woowacourse/react-shopping-cart/pull/357#discussion_r2116066573)`useEffect`를 인위적으로 트리거하여 리렌더링을 유발하는 형태였습니다.

이 `refetch` 기능은 수량 조절이나 삭제 등의 여러 기능에서 필요하며, SSOT 원칙을 유지하고 props drilling 문제를 방지하기 위해 별도의 `cartItems` Context를 만들었습니다. 

이를 통해 처음 PR 본문에 있었던 아래 고민도 해소할 수 있었습니다.

> Context API로 관리할 상태
a. selectedCartItems
단점: 선택되지 않은 cartItems는 전역적으로 관리되지 않습니다
> 

저는 처음에 SSOT를 지키려면 하나의 Context API로만 관리해야하는 줄 알았습니다. 하지만 실제로는 정말 별개의 개념이었으며, 이를 깨닫고 별도의 Context로 나누어 관리했습니다.

아무튼 단일원천을 지키기 위해 다음과 같이 접근했습니다.

- **CartItems Context**: 서버에서 관리되는 상태 원천
- **SelectedItems Context**: 클라이언트에서 관리되는 상태 원천

해당 내용에 관한 수정은 아래 커밋에 다 진행했습니다.

[refactor: 전반적인 상태 관리 구조 변경](https://github.com/woowacourse/react-shopping-cart/pull/357/commits/f3ecb00f8c6fd7501f672b6aa019eb21add3803c)

### 의문점 및 질문

두 개의 원천(`CartItems`, `SelectedItems`)을 따로 두었지만, 이것이 정말로 SSOT 원칙에 부합하는지 약간의 의문이 남습니다.

예를들어, `SelectedItems`에서 한 데이터를 제거해야하는 상황이 발생한다면, 동기화 관점에서 CartItems에서 해당 데이터도 제거가 돼야하지 않나? 라는 생각이 들었어요.

근데 이건 `SelectedItems` 데이터를 제거하는 로직에 연관된 `CartItems`도 제거하도록 구현하면 될 것 같은데, 이게 올바른 접근 방식인가? 라는 의문이 들었습니다.

구현이 누락될 경우 두 데이터 간의 동기화 문제가 생길 수 있으며, 이는 결국 SSOT 원칙을 위배한다고 생각했는데, 이에 대한 도밥의 생각을 듣고 싶습니다.

결국 저는 ‘앞서 말한 구현의 문제가 없다면 SSOT가 맞지 않을까?’ 라는 생각에 이렇게 진행하긴 했는데 아직 의문은 드는 점이라 질문드립니다!

### 고민했지만 반영하지 않은 부분

`cartItems`, `selectedCartItems`의 각 Context Provider에 있는 코드들을 각각 커스텀훅으로 분리하고, 하나의 Context로 관리할까 생각했습니다.

현재 `CartItemsProvider`에는 장바구니 아이템을 삭제하는 함수는 없고, 해당 기능을 사용하는 `CartItemCart.tsx`에 `handleCartItemDelete`로 있는데요, 

이는 추후 다른 페이지에서도 삭제 기능이 필요할때를 대비해서 `CartItemsProvider`에 추가하려 했습니다. 그러면 클라이언트 상태에서도 업데이트를 해줘야하니 `SelectedCartItemsProvider`에 있는 `removeSelectedCartItem`를 `CartItemsProvider`에 추가했었는데요, 

이미 `SelectedCartItemsProvider`에는 `CartItemsProvider`에서 `cartItems`를 불러와 사용하고 있기에 **context API 서로 의존하는 문제(순환 의존성)가 발생**해서 애초에 이렇게 로직을 작성할 수가 없었습니다.

그래서 이걸 어떻게 해결하지 고민하다가 하나의 Context API로 합치는 방법만 떠올랐습니다.

하지만.. 시간관계상, 그리고 그 과정에서 어떤 난관이 펼쳐질까 두렵고.. step2에서는 로컬 스토리지를 쓴다기에 그럼 Context API를 쓸 필요가 없어지는 건 아닐까?(같은 전역 관리 느낌이므로) 

여러 고민을 통해, 지금 미션에서는 그냥 현재처럼 냅둬도 문제는 없기에 반영하진 않았습니다.

Q.
비록 반영하진 않았지만 제가 생각한 문제 해결 방향성이 적절한 방향성일까요?
아니면 도밥이라면 어떻게 접근하실건가요?

Creative-Lee commented on .github/pull_request_template.md:
.

Creative-Lee commented on src/shared/context/SelectedCartItemsProvider.tsx:
```diff
@@ -15,26 +16,24 @@ interface SelectedCartItemsProviderProps {
-      return [...prevItems, cartItem];
-    });
+  useEffect(() => {
+    setSelectedCartItemsItems((prev) =>
```
`[SelectedCartItemsItems, setSelectedCartItemsItems]`

네이밍에 중복이 있는것 같아용

sanghee01 replied:
어맛 수정하겠습니다 .. 👍

Creative-Lee reviewed (approved):
안녕하세요 상추 ㅎㅎ
많은 코멘트 반영하신다고 고생 많으셨습니다.

---

질문에 대해 답변 남겨보자면, 

> 두 개의 원천(CartItems, SelectedItems)을 따로 두었지만, 이것이 정말로 SSOT 원칙에 부합하는지 약간의 의문이 남습니다.

CartItems과 SelectedItems 모두 장바구니 상품 api 라는 하나의 데이터 원천을 기반으로 생성되었으니 부합한다고 할 수 있을 것 같아요 ㅎㅎ
('두개의 원천'이라는 표현은 틀린것 같고, 2개의 상태가 하나의 원천으로 만들어졌다.라고 말해야할 것 같네요)

> 예를들어, SelectedItems에서 한 데이터를 제거해야하는 상황이 발생한다면, 동기화 관점에서 CartItems에서 해당 데이터도 제거가 돼야하지 않나? 라는 생각이 들었어요.
근데 이건 SelectedItems 데이터를 제거하는 로직에 연관된 CartItems도 제거하도록 구현하면 될 것 같은데, 이게 올바른 접근 방식인가? 라는 의문이 들었습니다.

선택된 장바구니 상태에서 '데이터를 제거해야하는 상황'이 정확이 어떤말인지 이해가 어렵네요~!
checkBox 토글을 통해 선택된 장바구니 상품 상태를 수정해야하는 것을 말하는게 맞나요? (=`removeSelectedCartItems` 함수)
맞다면...
`선택된 장바구니 상품 상태` 는 클라이언트에서 관리하는 상태입니다.(그 원천은 api)
선택하지 않은 상태로 만드는 행위는 실제 서버 데이터에 영향을 주는 행위가 아니며, cartItems 상태(서버에서 가져온 데이터를 넣어놓은 상태)를 제거하는것과는 관련이 없는것 같아요.

한편 서버 데이터의 변화를 주는 행위, 예를들어 장바구니에서 삭제를 하는 경우에는 말이 다르겠죠.
장바구니 상품 삭제 api 호출 -> 장바구니 get api 호출로 데이터 최신화 -> 최신화된 데이터 기반으로 `선택된 장바구니 상품` 상태 최신화 
등의 행위를 해야합니다. (원천 데이터가 변경되었으니, 이를 기반으로 만든 `선택된 장바구니 상품`를 변경 하는거죠) 

혹은 자신이 설계한 방향에 따라 장바구니 get api 호출을 활용한 데이터 최신화 대신,
장바구니 상품 삭제 api 호출 -> 삭제 성공 여부에 따라 `장바구니 상품 상태` 및 `선택된 장바구니 상품샹태`를 직접 최신화 (새로고침 시에만 get api 호출 및 상태 설정하는 개념)
등의 행위를 해야할수도 있죠. 

위에서 설명한 방식들은 모두 SSOT를 지켰습니다. 상태 동기화와 관련된 방식이 다른것 뿐입니다.
그래서 상추가 어떤 의도로 어떤 설계를 했느냐를 잘 정리해보는게 중요합니다.
(설계가 변경된것은 알겠는데, 상태 설계나 동기화 흐름에 대해서는 저에게 따로 말씀주시지 않았어요! 요것도 다음 PR에 함번 담아주세요 ㅎㅎ)

---

고민했지만 반영하지 않은 부분에 대해서는

`장바구니 상품 목록 상태` 와 `선택된 상품 목록 상태`는 원천은 같지만, 필요로하는 이유는 명확하게 다릅니다.
관심사 측면에서는 
선택된 상품 목록 상태를 다루는 provider에서 필요한 동작은 '목록에 상품을 넣기/빼기' 입니다. 
상품 목록 상태를 다루는 provider에서 필요한 동작은 '상품 삭제', '수량 변경' 입니다.
이런 맥락에서 호출처(컴포넌트)에서 각 로직을 나열해서 작성하는 것은 크게 어색하게 보이진 않습니다.
context를 합치는 부분에서는
둘을 하나로 합친다는 개념보다는 둘 모두를 호출해서 최종 action을 return 하는 훅을 만드는 방식이 더 나아보입니다.

---

고생 많으셨습니다~ 상추 👍 
step2를 진행하면서 생각의 변화가 있을만한 부분이 있을것 같아서 다음으로 넘어가보는게 좋겠습니다 ㅎㅎ
(혹시 이해가지 않거나 더 궁금한 부분이 있다면, DM 주세요! 추가 설명이나 허들등으로 같이 해결해봐요~!)
step2에서 뵐게요💪
